### PR TITLE
Issue #5818: Try to migrate default search engine.

### PR DESCRIPTION
**Pull request by pocmo**
Implementation for migrating the default search engine from Fennec to Fenix.

This is a "best effort" migration. We will recover the default search engine name from Fennec and then try to find a matching search engine in the search engine manager for the current Fenix installation. If we find a match we will update the default search engine to that; otherwise the regular default search engine for this installation will be used.

### CODE REVIEW

File: components/support/migration/src/main/java/mozilla/components/support/migration/TelemetryHelpers.kt [hunk 1]
```diff
@@ -16,6 +16,7 @@ internal fun Migration.telemetryIdentifier(): String {
         Migration.Settings -> "settings"
         Migration.Addons -> "addons"
         Migration.TelemetryIdentifiers -> "telemetry_identifiers"
+        Migration.SearchEngine -> "search_engine"
```
Comment (grigoryk) [inline components/support/migration/src/main/java/mozilla/components/support/migration/TelemetryHelpers.kt]: in metrics.yaml `migration_versions` you called it `search` - these labels need to be the same (also, see comment on top of this function).

File: components/support/migration/src/main/java/mozilla/components/support/migration/SearchEngineMigration.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    private suspend fun determineNewDefault(
+        context: Context,
+        searchEngineManager: SearchEngineManager,
+        name: String
+    ): Result<SearchEngineMigrationResult> {
+        val searchEngines = searchEngineManager.getSearchEnginesAsync(context)
+
+        searchEngines.forEach { engine ->
```
Comment (grigoryk) [inline components/support/migration/src/main/java/mozilla/components/support/migration/SearchEngineMigration.kt]: maybe a comment here would be useful - if passed-in fennec default doesn't match anything from the search manager, we do nothing - which is equivalent of keeping fenix's default search engine as a "fallback".

Comment (pocmo) [reply components/support/migration/src/main/java/mozilla/components/support/migration/SearchEngineMigration.kt]: That is mentioned in the class comment. But I agree, here in context, we could mention that too.

**Comment by pocmo** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

This is an extension of the already existing migration telemetry ([data review](https://github.com/mozilla-mobile/android-components/pull/5483#issuecomment-573971458)). This patch set extends the telemetry probes to cover the new "default search engine" migration, with the goal of measuring the success/error rates of this new migration type.

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?

This information is essential to ensure the default search engine migration as part of the Fennec->Fenix migration proceeds without unexpected errors. The information is used to determine the overall success rate and to decide whether additional work is needed to achieve a higher success rate.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Crash and failed assertion reporting via Socorro and Sentry is in place to debug individual errors, but that isn't detailed enough for us to have a good picture of quality of the migrations in the wild.

4) Can current instrumentation answer these questions?

No.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Search engine migration</td>
    <td>Category 2, Interaction data</td>
    <td>https://github.com/mozilla-mobile/fenix/issues/5818</td>
  </tr>
</table>

In detail:

* We add an additional label to the `migration_versions` probe to collect the version of the search engine migration we ran (`search` / `labeled_string`)
* We add `migration.search.any_failures` (`boolean`) to indicate whether failures occured while migrating the search engine.
* For failures we add `migration.search.failure_reason` to indicate the reason why the migration failed (one out of three different constants)
* For successful migrations we add `migration.success_reason` to indicate why/how the migration succeeded (currently only one constant)

6) How long will this data be collected?  Choose one of the following:

For the same time as the other migration probes (see [data review](https://github.com/mozilla-mobile/android-components/pull/5483#issuecomment-573971458)).

7) What populations will you measure?

* Which release channels?
all

* Which countries?
all

* Which locales?
all

* Any other filters?  Please describe in detail below.
no

8) If this data collection is default on, what is the opt-out mechanism for users?
No opt-out mechanism.

9) Please provide a general description of how you will analyze this data.
Via the migration dashboard on sql.telemetry.mozilla.org.
 
10) Where do you intend to share the results of your analysis?
Internally within the core Fenix/A-C/A-S teams.

11) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:

n/a

File: components/support/migration/src/main/java/mozilla/components/support/migration/TelemetryHelpers.kt (hunk 1 above)
Comment (pocmo) [reply components/support/migration/src/main/java/mozilla/components/support/migration/TelemetryHelpers.kt]: Argh, wrote that before any other telemetry code. :)

**Comment by pocmo** [comment]
Added some tests. I think this is now ready to land after the data review.

**Pull request merged**